draft-peabody-dispatch-new-uuid-format-03.txt

```diff
@@ -321,1105 +322,694 @@ Internet-Draft               new-uuid-format               February 2022
+   9 to 64 bytes.  (TODO: decide if we want to elaborate on the various
+   possible motivations for this here, e.g. shorter for better human
+   use, longer for lower collision probability)
 
+   Implementations that store UUIDs of variable length SHOULD support
+   any length from 9 to 64 bytes (TODO: decide if this is the right
+   range - less than 9 bytes and we lose the variant/version, and 64 is
+   arbitrary but we should have some upper limit in case it has an
+   impact on physical storage requirements)
 
 
 
 
 
-
-
-Peabody & Davis           Expires 5 August 2022                [Page 20]
+Peabody & Davis          Expires 20 August 2022                [Page 13]
 
 Internet-Draft               new-uuid-format               February 2022
 
 
-   The clock sequence MUST start at zero and increment monotonically for
-   each new UUIDv8 created on by the application on the same timestamp.
-   When the timestamp increments the clock sequence MUST be reset to
-   zero.  The clock sequence MUST NOT rollover or reset to zero unless
-   the timestamp has incremented.  Care MUST be given to ensure that an
-   adequate sized clock sequence is selected for a given application
-   based on expected timestamp precision and expected UUIDv8 generation
-   rates.
+   Variable length is optional and implementations MAY opt-out and only
+   support 128-bits.
 
-4.5.3.  UUIDv8 Node Usage
+   Recommendation is for implementations to start defining UUID as an
+   array of bytes (length + pointer or whatever language mechanism)
+   instead of a fixed set of 16 bytes.
 
-   The UUIDv8 Node MAY contain any set of data an implementation desires
-   however the node MUST NOT be set to all 0s which does not ensure
-   global uniqueness.  In most scenarios the node will be filled with
-   pseudo-random data.
+   TODO: I think we need a table in here that gives bit/byte lengths and
+   collision probability.  So someone can look down the numbers and pick
+   the one that is right for their app.
 
-   The UUIDv8 layout in Figure 6 defines 2 sizes of Node depending on
-   the timestamp size:
+7.4.  Text Format
 
-   *  62 bit node encompassing seq_or_node and node Used when a
-      timestamp of 48 bits or less is leveraged.
-   *  54 bit node when all 60 bits of the timestamp are in use and the
-      seq_or_node is used as clock sequencing.
+   The existing hex and dash format (8-4-4-4-12) is still valid.
+   Shorter or longer values just omit or add hex characters (TODO: give
+   some examples at different lengths)
 
-   An implementation MAY choose to allocate bits from the node to the
-   timestamp, clock sequence or application-specific embedded field.  It
-   is recommended that implementation utilize a node of at least 48 bits
-   to ensure global uniqueness can be guaranteed.
+   Crockford base32 is allowed also (TODO: briefly summarize reasons -
+   shorter, more compact if UUIDs are stored as text, crockford
+   version's benefits vs).
```

Comment:
I have another question about encoding:

    how to encode a UUID to base-32?
    should I encode using 40-bit groups, from left to right, as in [RFC-4648, section 6](https://datatracker.ietf.org/doc/html/rfc4648#section-6)?
    or should I code using a generic algorithm that performs [modulo division](https://cs.stackexchange.com/questions/10318/the-math-behind-converting-from-any-base-to-any-base-no-cross-base)?

The RFC-4648 algorithm is much faster than the modulus division algorithm, but it produces a completely different result. To produce the same result with RFC-4648 algorithm, it is necessary to pad the low-order bits with zeros.

There are actually 3 questions :)

Reply:
The RFC-4648 algorithm will be the only way to go if the variable length fields are to be added at the bottom, though that is different than ULID's encoding.

But anyway Crockford base32's tolerance over 0o1il breaks ASCII-sortability of the textual representations. This fact effectively requires all the implementations to canonicalize input before sorting, which considerably deteriorates the usefulness of the new spec. Lexicographical sortability should be the first priority of the new spec, so the letter/digit errors should be strictly prohibited at the spec level. IMO, Crockford base32's human-friendliness is a sort of area that should be covered by check digit algorithms.

Plus, though Base 32 is convenient in that it is case-insensitive and url-safe, it is still redundant compared to Base 64 schemes. Further research on needs and use cases will be necessary before standardization.

Reply:
Yes. The RFC-4648 is the only way for variable length UUID.

As for the Crockford alphabet, months ago I thought it would be a good choice. But now I'm not sure.

Reply:
Another thing we haven't thought about yet is what the URN representation for base 32 encoded UUIDs will look like. 

`urn:uuid:01BX5ZZKBKACTAV9WEVGEMMVRY`

Would this format above be correct?

Reply:
One more thing: the Crockford alphabet is encoded with UPPER CASE, just like the alphabet in RFC-4648.

Will the base 32 UUIDs be uppercase encoded as well? Either way this needs to be present in the draft.

Reply:
Yet another: the [formula](https://github.com/f4b6a3/uuid-creator/blob/dbc54155c6cb2893eff99876b75ead20382fe331/src/main/java/com/github/f4b6a3/uuid/codec/base/BaseN.java#L134) to calculate the size of a variable-length UUID in base-32 must be included in the draft. We know that a base 32 encoded 128-bit UUID has 22 characters (without padding), but we will have to calculate this for each UUID length.

Reply:
The check character at the end of a Crockford-encoded UUID could be a problem with the variable-length UUID. Is the last character part of the UUID or just a check symbol?

I'm commenting as soon as I remember and before I forget. Sorry. :)

Reply:
I wouldn't say a UUID should have a check digit, but at the same time I wouldn't think a UUID needs to be as human-error-tolerant as Crockford base 32. The best textual encoding should be determined based on real needs, but I don't clearly see such needs as of now. That is my concern.

We should also be aware of other attempts like https://datatracker.ietf.org/doc/draft-taylor-uuid-ncname/ . Or we can just fall back on a approach like https://datatracker.ietf.org/doc/draft-multiformats-multibase/ .

Reply:
This thread is getting a bit long in the tooth. Should we move discussion to #2 ?

Reply:
Thanks group, I re-opened that thread.